nifi-nar-bundles/nifi-elasticsearch-bundle/nifi-elasticsearch-restapi-processors/src/main/java/org/apache/nifi/processors/elasticsearch/PutElasticsearchRecord.java

```diff
@@ -0,0 +1,254 @@
+    static final PropertyDescriptor RECORD_READER = new PropertyDescriptor.Builder()
+        .name("put-es-record-reader")
+        .displayName("Record Reader")
+        .description("The record reader to use for reading incoming records from flowfiles.")
+        .identifiesControllerService(RecordReaderFactory.class)
+        .required(true)
+        .addValidator(Validator.VALID)
+        .build();
+
+    static final PropertyDescriptor OPERATION_RECORD_PATH = new PropertyDescriptor.Builder()
+        .name("put-es-record-operation-path")
+        .displayName("A record path expression to retrieve index operation setting from each record. If left blank, " +
+                "all operations will be assumed to be index operations.")
+        .addValidator(Validator.VALID)
+        .required(false)
+        .expressionLanguageSupported(ExpressionLanguageScope.FLOWFILE_ATTRIBUTES)
+        .build();
+
+    static final PropertyDescriptor ID_RECORD_PATH = new PropertyDescriptor.Builder()
+        .name("put-es-record-id-path")
+        .displayName("ID Record Path")
+        .description("A record path expression to retrieve the ID field for use with ElasticSearch. If left blank " +
+                "the ID will be automatically generated by ElasticSearch.")
+        .addValidator(Validator.VALID)
+        .expressionLanguageSupported(ExpressionLanguageScope.FLOWFILE_ATTRIBUTES)
+        .build();
+
+    static final PropertyDescriptor INDEX_RECORD_PATH = new PropertyDescriptor.Builder()
+        .name("put-es-record-index-record-path")
+        .displayName("Index Record Path")
+        .description("A record path expression to retrieve the ID field for use with ElasticSearch. If left blank " +
+                "the index will be determined using the main index property.")
+        .addValidator(Validator.VALID)
+        .required(false)
+        .expressionLanguageSupported(ExpressionLanguageScope.FLOWFILE_ATTRIBUTES)
+        .build();
+
+    static final PropertyDescriptor TYPE_RECORD_PATH = new PropertyDescriptor.Builder()
+        .name("put-es-record-type-record-path")
+        .displayName("Type Record Path")
+        .description("A record path expression to retrieve the type field for use with ElasticSearch. If left blank " +
+                "the type will be determined using the main type property.")
+        .addValidator(Validator.VALID)
+        .required(false)
+        .expressionLanguageSupported(ExpressionLanguageScope.FLOWFILE_ATTRIBUTES)
+        .build();
+
+    @Override
+    public Set<Relationship> getRelationships() {
+        return new HashSet<Relationship>(){{
+            add(REL_SUCCESS);
+            add(REL_FAILURE);
+            add(REL_RETRY);
+        }};
+    }
+
+    @Override
+    public final List<PropertyDescriptor> getSupportedPropertyDescriptors() {
+        List<PropertyDescriptor> _temp = new ArrayList<>();
+        _temp.add(INDEX);
```

Comment:
Can't remember if I made the comment here or in another PR, but you could always add these properties in the constructor or static block so the list only gets created once. I don't think it creates a memory leak here, so I guess this is just a style thing